File: ebike-service/src/main/java/it/unibo/sap/ass02/demo/domain/EBike.java
package it.unibo.sap.ass02.demo.domain;


import lombok.Data;

public interface EBike<K> extends Entity<K>{
    EBikeState getState();

    void rechargeBattery();

    int getBatteryLevel();

    void decreaseBatteryLevel(int delta);

    boolean isAvailable();

    void updateState(EBikeState state);

    void updateLocation(P2d newLoc);

    void updateSpeed(double speed);

    void updateDirection(V2d dir);

    double getSpeed();

    V2d getDirection();

    P2d getLocation();

    enum EBikeState {
        AVAILABLE,
        IN_USE,
        MAINTENANCE;

        public static EBikeState fromValue(String value) {
            // Normalize the string (e.g., lowercase) and map it to the enum
            return switch (value.toUpperCase()) {
                case "AVAILABLE" -> AVAILABLE;
                case "IN_USE" -> IN_USE;
                case "MAINTENANCE" -> MAINTENANCE;
                default -> throw new IllegalArgumentException("Unknown value: " + value);
            };
        }
    }
}

File: ride-service/core/src/main/kotlin/it/unibo/sap/ass02/domain/model/EBike.kt
package it.unibo.sap.ass02.domain.model

import it.unibo.sap.ass02.domain.ddd.Entity

interface EBike : Entity<String> {
    var location: P2d
    var direction: V2d
    var speed: Double
    var state: String // TODO: create an enum
    var available: Boolean
    var battery: Int

    fun updateSpeed(speed: Double)

    fun updateLocation(pos: P2d)

    fun updateDirection(dir: V2d)

    fun drainBattery(delta: Int = 1)
}
